crates/sui-indexer-alt/src/handlers/kv_epoch_ends.rs

```diff
@@ -0,0 +1,132 @@
+            )
+        }) else {
+            bail!(
+                "Failed to get end of epoch transaction in checkpoint {} with EndOfEpochData",
+                checkpoint_summary.sequence_number,
+            );
+        };
+
+        if let Some(SystemEpochInfoEvent {
+            total_stake,
+            storage_fund_reinvestment,
+            storage_charge,
+            storage_rebate,
+            storage_fund_balance,
+            stake_subsidy_amount,
+            total_gas_fees,
+            total_stake_rewards_distributed,
+            leftover_storage_fund_inflow,
+            ..
+        }) = transaction
+            .events
+            .iter()
+            .flat_map(|events| &events.data)
+            .find_map(|event| {
+                event
+                    .is_system_epoch_info_event()
+                    .then(|| bcs::from_bytes(&event.contents))
+            })
+            .transpose()
+            .context("Failed to deserialize SystemEpochInfoEvent")?
+        {
+            Ok(vec![StoredEpochEnd {
+                epoch: checkpoint_summary.epoch as i64,
+                cp_hi: checkpoint_summary.sequence_number as i64 + 1,
+                tx_hi: checkpoint_summary.network_total_transactions as i64,
+                end_timestamp_ms: checkpoint_summary.timestamp_ms as i64,
+
+                safe_mode: false,
+
+                total_stake: Some(total_stake as i64),
+                storage_fund_balance: Some(storage_fund_balance as i64),
+                storage_fund_reinvestment: Some(storage_fund_reinvestment as i64),
+                storage_charge: Some(storage_charge as i64),
+                storage_rebate: Some(storage_rebate as i64),
+                stake_subsidy_amount: Some(stake_subsidy_amount as i64),
+                total_gas_fees: Some(total_gas_fees as i64),
+                total_stake_rewards_distributed: Some(total_stake_rewards_distributed as i64),
+                leftover_storage_fund_inflow: Some(leftover_storage_fund_inflow as i64),
+
+                epoch_commitments: bcs::to_bytes(&end_of_epoch.epoch_commitments)
+                    .context("Failed to serialize EpochCommitment-s")?,
+            }])
+        } else {
+            Ok(vec![StoredEpochEnd {
+                epoch: checkpoint_summary.epoch as i64,
+                cp_hi: checkpoint_summary.sequence_number as i64 + 1,
+                tx_hi: checkpoint_summary.network_total_transactions as i64,
+                end_timestamp_ms: checkpoint_summary.timestamp_ms as i64,
+
+                safe_mode: true,
+
+                total_stake: None,
+                storage_fund_balance: None,
+                storage_fund_reinvestment: None,
+                storage_charge: None,
+                storage_rebate: None,
+                stake_subsidy_amount: None,
+                total_gas_fees: None,
+                total_stake_rewards_distributed: None,
+                leftover_storage_fund_inflow: None,
```

Comment:
how about a default or something

Reply:
There is no such thing as a "Default" ending of an epoch -- we shouldn't use `Default` (only) because we're tired of writing out fields, the type needs to have a reasonable notion of a "default value".